---
title: "NaRnEA - User Guide Part 1"
output: rmarkdown::html_vignette
theme: united
highlight: tango
vignette: >
  %\VignetteIndexEntry{NaRnEA - User Guide Part 1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
opts_chunk$set(collapse = TRUE, comment = "#>", tidy = TRUE, tidy.opts=list(width.cutoff=70))
```

## Introduction

Non-parametric analytical Rank-based Enrichment Analysis (NaRnEA) is a novel statistical method created by [Griffin et al.](https://www.biorxiv.org/search/NaRnEA) for the purpose of performing powerful and accurate gene set analysis. 

## Installing NaRnEA

The `NaRnEA` R package can be installed from the [Califano Lab Github](https://github.com/califano-lab/NaRnEA) using the `devtools` R package with the following code.
```{r, eval = FALSE}
devtools::install_github(repo = "califano-lab/NaRnEA", force = TRUE, build_vignettes = TRUE)
```

## NaRnEA Statistical Theory

To learn more about the statistical theory behind NaRnEA, check out the "NaRnEA - Statistical Theory" vignette with the following command.
```{r, eval = FALSE}
vignette(topic = "NaRnEA_Theory", package = "NaRnEA")
```

## Simulated Gene Set Analysis with NaRnEA

We'll begin our user guide with something very simple, demonstrating the enrichment of a gene set in a simulated gene expression signature at varying levels of enrichment which we will manually control. This will allow us to observe the behavior of NaRnEA in a controlled setting. First let's load `NaRnEA` as well as `ggplot2`, a `tidyverse` package that we'll use for plotting.
```{r, eval = FALSE}
library(NaRnEA)
library(ggplot2)
```

Next we'll set the seed for the simulation, which will ensure that our simulated gene set and simulated gene expression signature are reproducible.
```{r, eval = FALSE}
sim.seed <- 11
set.seed(sim.seed)
```

Now let's simulate a gene expression signature; we can do this in a very simple manner by simulating gene expression signature values for 20,000 genes uniformly on the interval [-10,10]. The gene expression signature vector should also be named so that we know which value corresponds to which gene.
```{r, eval = FALSE}
gene.num <- 20000
null.ges.min <- -10
null.ges.max <- 10
synthetic.ges <- runif(n = gene.num, min = null.ges.min, max = null.ges.max)
names(synthetic.ges) <- paste("g",1:length(synthetic.ges),"",sep = "_")
```

Now let's create a gene set by randomly selecting some genes from the gene expression signature; we'll go with 100 genes for now. Each gene in the gene set will need to be parameterized with a Mode of Regulation (tfmode) and a Regulation Confidence (likelihood) parmeter. NaRnEA will use these parameters to determine how the differential expression of each gene contribtes to the enrichment of the gene set in the gene expression signature
```{r, eval = FALSE}
gene.set.size <- 100
gene.set.MoR.min <- -1
gene.set.MoR.max <- 1
gene.set.RC.min <- 0
gene.set.RC.max <- 1

synthetic.gene.set.targets <- sample(names(synthetic.ges), size = gene.set.size)
synthetic.gene.set <- list(tfmode = runif(n = gene.set.size, min = gene.set.MoR.min, max = gene.set.MoR.max), likelihood = runif(n = gene.set.size, min = gene.set.RC.min, max = gene.set.RC.max))
names(synthetic.gene.set$tfmode) <- synthetic.gene.set.targets
```

Because the members of this gene set were selected at random from the gene expression signature we know the null hypothesis of gene set analysis is true for this gene set (i.e. the gene set is not enriched in the gene expression signature). Let's check to see if NaRnEA agrees with this ground truth; we can test this numerically by calling the NaRnEA function directly or by using the visualization function EnrichmentAnalysisPlot.
```{r, eval = FALSE}
NaRnEA(ges = synthetic.ges, regulon = synthetic.gene.set)

EnrichmentAnalysisPlot(cur.ges = synthetic.ges, cur.regul = synthetic.gene.set)
```
```{r, echo = FALSE}
knitr::include_graphics(path = "../man/figures/NaRnEA_User_Guide_1_Fig_1.png", dpi = 500)
```

The EnrichmentAnalysisPlot function was designed to allow users to get a sense of where the genes in the gene set are distributed in the gene expression signature in a manner that is sensitive to the Mode of Regulation and Regulation Confidence while returning the NaRnEA proportional enrichment score, normalized enrichment score, and two-sided p-value. Additionally, the EnrichmentAnalysisPlot function allows users to get a sense of the enrichment that earlier gene set analysis methods, like the two-sample Kolmogorov-Smirnov (KS) test or Gene Set Enrichment Analysis (GSEA), would have inferred for the gene set.

We see from this analysis that the NaRnEA proportional enrichment score is close to zero, the normalized enrichment score has a small magnitude, and the two-sided p-value is not significant. Now let's modify the gene expression signature values for the genes in the gene set in order to explore the behavior of NaRnEA when the gene set is non-uniformly distributed in the gene expression signature
```{r, eval = FALSE}
alt.ges.min <- 0
alt.ges.max <- 10
synthetic.ges[match(names(synthetic.gene.set$tfmode),names(synthetic.ges))] <- runif(n = length(synthetic.gene.set$tfmode), min = alt.ges.min, max = alt.ges.max)*sign(synthetic.gene.set$tfmode)

NaRnEA(ges = synthetic.ges, regulon = synthetic.gene.set)

EnrichmentAnalysisPlot(cur.ges = synthetic.ges, cur.regul = synthetic.gene.set)
```
```{r, echo = FALSE}
knitr::include_graphics(path = "../man/figures/NaRnEA_User_Guide_1_Fig_2.png", dpi = 500)
```

We've simulated a new value for each gene in the gene set in agreement with each gene's Mode of Regulation which corresponds with the positive alternative hypothesis of gene set analysis. Based on the NaRnEA two-sided p-value we correctly reject the null hypothesis in favor of the positive alternative hypothesis based on the sign of the proportional enrichment score. Now let's steadily increase the enrichment of the gene set and see how NaRnEA picks up on this.
```{r, eval = FALSE}
alt.ges.min <- 2
alt.ges.max <- 10
synthetic.ges[match(names(synthetic.gene.set$tfmode),names(synthetic.ges))] <- runif(n = length(synthetic.gene.set$tfmode), min = alt.ges.min, max = alt.ges.max)*sign(synthetic.gene.set$tfmode)

NaRnEA(ges = synthetic.ges, regulon = synthetic.gene.set)

EnrichmentAnalysisPlot(cur.ges = synthetic.ges, cur.regul = synthetic.gene.set)
```
```{r, echo = FALSE}
knitr::include_graphics(path = "../man/figures/NaRnEA_User_Guide_1_Fig_3.png", dpi = 500)
```

```{r, eval = FALSE}
alt.ges.min <- 4
alt.ges.max <- 10
synthetic.ges[match(names(synthetic.gene.set$tfmode),names(synthetic.ges))] <- runif(n = length(synthetic.gene.set$tfmode), min = alt.ges.min, max = alt.ges.max)*sign(synthetic.gene.set$tfmode)

NaRnEA(ges = synthetic.ges, regulon = synthetic.gene.set)

EnrichmentAnalysisPlot(cur.ges = synthetic.ges, cur.regul = synthetic.gene.set)
```
ggsave(filename = "NaRnEA_User_Guide_1_Fig_4.png", device = "png", dpi = 500, path = "PNG_Figures")

```{r, eval = FALSE}
alt.ges.min <- 6
alt.ges.max <- 10
synthetic.ges[match(names(synthetic.gene.set$tfmode),names(synthetic.ges))] <- runif(n = length(synthetic.gene.set$tfmode), min = alt.ges.min, max = alt.ges.max)*sign(synthetic.gene.set$tfmode)

NaRnEA(ges = synthetic.ges, regulon = synthetic.gene.set)

EnrichmentAnalysisPlot(cur.ges = synthetic.ges, cur.regul = synthetic.gene.set)
```
```{r, echo = FALSE}
knitr::include_graphics(path = "../man/figures/NaRnEA_User_Guide_1_Fig_4.png", dpi = 500)
```

```{r, eval = FALSE}
alt.ges.min <- 8
alt.ges.max <- 10
synthetic.ges[match(names(synthetic.gene.set$tfmode),names(synthetic.ges))] <- runif(n = length(synthetic.gene.set$tfmode), min = alt.ges.min, max = alt.ges.max)*sign(synthetic.gene.set$tfmode)

NaRnEA(ges = synthetic.ges, regulon = synthetic.gene.set)

EnrichmentAnalysisPlot(cur.ges = synthetic.ges, cur.regul = synthetic.gene.set)
```
```{r, echo = FALSE}
knitr::include_graphics(path = "../man/figures/NaRnEA_User_Guide_1_Fig_5.png", dpi = 500)
```










