---
title: "NaRnEA User Guide - Part 2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{NaRnEA User Guide - Part 2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, include = FALSE}
library(knitr)
opts_chunk$set(collapse = TRUE, comment = "#>", tidy = TRUE, tidy.opts=list(width.cutoff=70))
```

## Introduction

Non-parametric analytical Rank-based Enrichment Analysis (`NaRnEA`) is a novel statistical method created by Griffin et al. in 2020 for the purpose of performing powerful and accurate gene set analysis. To read more about the theory behind NaRnEA and how it compares with Gene Set Enrichment Analysis (GSEA), which was created by Subramanian et al. in 2005, check out the preprint of the NaRnEA manuscript on bioRxiv ([link](https://www.biorxiv.org/search/NaRnEA)).

### Installing NaRnEA

For these analyses we'll need to install and load several R packages, including `NaRnEA`. For all of our visualization we'll be using the `ggplot2` package ([link](https://ggplot2.tidyverse.org/)), created by Hadley Wickham as part of the tidyverse. 

```{r, eval = FALSE}
install.packages("BiocManager")
library(BiocManager)

install("ggplot2")
library(ggplot2)

install("devtools")
library(devtools)

install_github(repo = "califano-lab/NaRnEA", force = TRUE, build_vignettes = TRUE)
library(NaRnEA)
```














```{r, eval = FALSE}





# set parameters
sim.seed <- 11
sig.figs <- 3
figure.output.value <- 2

# set the seed
set.seed(sim.seed)

# load in the data frame that will let us convert entrez id gene names into gene symbols, which was also the data frame used to convert the kallisto transcript expression profiles into gene expression profiles
entrez.to.hugo.data <- readRDS(file = "/Volumes/ATG_4TB_HD_v3/ATG_NARNEA_Project/110420/vignette_data/tx2gene.rds")

# load in the TCGA LUAD gene expression profiles (raw counts)
luad.counts <- readRDS(file = "/Volumes/ATG_4TB_HD_v3/ATG_NARNEA_Project/110420/vignette_data/TCGA_LUAD_RawCounts.rds")

# extract 50 tumor gene expression profiles and 50 tissue gene expression profiles from the total set of TCGA LUAD gene expression profiles
tumor.counts <- luad.counts[,which(substr(colnames(luad.counts),13,15) == "-01")]
tissue.counts <- luad.counts[,which(substr(colnames(luad.counts),13,15) == "-11")]

common.patients <- intersect(substr(colnames(tumor.counts),1,12),substr(colnames(tissue.counts),1,12))

tumor.counts <- tumor.counts[,match(common.patients[1:50],substr(colnames(tumor.counts),1,12))]
tissue.counts <- tissue.counts[,match(common.patients[1:50],substr(colnames(tissue.counts),1,12))]

# normalize the samples for sequencing depth
tumor.cpm <- apply(tumor.counts,2,function(x){
	y <- 1E6*x/sum(x)
	return(y)
})

tissue.cpm <- apply(tissue.counts,2,function(x){
	y <- 1E6*x/sum(x)
	return(y)
})

# combine the samples and compute a non-parametric internal gene expression signature for each sample
total.cpm <- cbind(tumor.cpm,tissue.cpm)

total.ges <- t(apply(total.cpm,1,function(x){
	y <- rank(x, ties.method = "random")/(1 + length(x))
	z <- qnorm(y)
	return(z)
}))

# compute a distance matrix between all samples based on Pearson Correlation
total.ges.dist <- sqrt((1 - cor(total.ges, method = "pearson")))
total.ges.dist <- as.dist(total.ges.dist)

# perform dimensionality reduction with the resultant distance matrix using multidemsional scaling
set.seed(sim.seed)

plot.data <- cmdscale(d = total.ges.dist, k = 2)[,1:2]
plot.data <- as.data.frame(plot.data)
colnames(plot.data) <- c("x.values","y.values")
plot.data$group.values <- c(rep("LUAD", times = 50), rep("LUNG", times = 50))
plot.data$group.values <- factor(plot.data$group.values, levels = c("LUAD","LUNG"))

colour.value.vec <- c("LUAD" = "purple", "LUNG" = "forestgreen")

x.step <- .1
x.max <- ceiling(max(abs(plot.data[,c("x.values","y.values")])/x.step))*x.step
x.min <- -1*x.max
x.breaks <- seq(from = x.min, to = x.max, by = x.step)

y.breaks <- x.breaks

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- "MDS Dimensionality Reduction"
cur.x.lab <- "Dimension 1"
cur.y.lab <- "Dimension 2"
cur.colour.lab <- "Phenotype"

ggplot(plot.data) + geom_hline(colour = "black", yintercept = 0, lwd = 1) + geom_vline(colour = "black", xintercept = 0, lwd = 1) + geom_point(aes(x = x.values, y = y.values, colour = group.values), shape = 21, fill = "white", size = 4, alpha = 1, stroke = 2) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + scale_colour_manual(values = colour.value.vec) + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# perform some unsupervised clustering on the gene expression profiles
total.ges.clust <- pam(x = total.ges.dist, k = 2)

# visualize the unsupervised clustering silhouette scores
plot.data <- as.data.frame(total.ges.clust$silinfo$widths)
colnames(plot.data) <- c("cluster.values","neighborh.values","score.values")
plot.data$group.values <- "LUAD"
plot.data[which(rownames(plot.data)%in%colnames(tissue.counts)),"group.values"] <- "LUNG"
plot.data$cluster.values <- paste("Cluster",as.character(plot.data$cluster.values),sep = " ")
plot.data$sample.values <- rownames(plot.data)
plot.data$sample.values <- factor(plot.data$sample.values, levels = rev(as.character(plot.data$sample.values)))

colour.value.vec <- c("LUAD" = "purple", "LUNG" = "forestgreen")

x.breaks <- seq(from = -.2, to = 1, by = .1)

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("PAM Unsupervised Clustering", "\n", "Fisher's Exact Test (Cluster vs. Phenotype)", "\n", "Estimated Odds Ratio = ", signif(fisher.test(table(plot.data[,c("cluster.values","group.values")]), alternative = "greater")$estimate, sig.figs), " (p = ",signif(fisher.test(table(plot.data[,c("cluster.values","group.values")]), alternative = "greater")$p.value, sig.figs),")", sep = "")
cur.x.lab <- "Silhouette Score"
cur.y.lab <- "Sample"
cur.colour.lab <- "Phenotype"

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = 0) + facet_wrap(~cluster.values, ncol = 1, scales = "free") + geom_segment(aes(x = score.values, xend = 0, y = sample.values, yend = sample.values, colour = group.values), lwd = 1.5) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_colour_manual(values = colour.value.vec) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 6), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom", strip.text = element_text(hjust = 0.5, colour = "black", size = 20)) + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")


# compute the differential expression of all protein coding genes between LUAD and LUNG using DESeq2
cur.count.data <- cbind(tumor.counts,tissue.counts)
cur.count.data <- apply(cur.count.data,2,function(x){
	y <- round(x)
	return(y)
})

cur.meta.data <- data.frame(Site = rep("TCGA", times = 100), Group = c(rep("Tumor", times = 50),rep("Tissue", times = 50)))
rownames(cur.meta.data) <- colnames(cur.count.data)

cur.dds <- DESeqDataSetFromMatrix(countData = cur.count.data, colData = cur.meta.data, design = ~Group)
cur.dds$Group <- relevel(cur.dds$Group, ref = "Tissue")

cur.dds <- DESeq(cur.dds)

cur.dge.res <- results(cur.dds, contrast = c("Group","Tumor","Tissue"), cooksCutoff = FALSE, independentFiltering = FALSE, format = "DataFrame")
cur.dge.res <- as.data.frame(cur.dge.res)

# visualize the differential gene expression results
plot.data <- data.frame(x.values = cur.dge.res$log2FoldChange, p.values = cur.dge.res$pvalue)
plot.data <- plot.data[which(is.finite(plot.data$p.values)),]
plot.data$y.values <- -1*log(plot.data$p.values,base = 10)
plot.data$fdr.values <- p.adjust(p = plot.data$p.values, method = "BH")
plot.data$dir.values <- "NONE"
plot.data[which(plot.data$x.values < 0 & plot.data$fdr.values < .05),"dir.values"] <- "DOWN"
plot.data[which(plot.data$x.values > 0 & plot.data$fdr.values < .05),"dir.values"] <- "UP"
plot.data$dir.values <- factor(plot.data$dir.values, levels = c("DOWN","NONE","UP"))

colour.value.vec <- c("DOWN" = "forestgreen", "NONE" = "grey50", "UP" = "purple")

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("DESeq2-Inferred Differential Gene Expression","\n","Downregulated Genes = ", sum(plot.data$dir.values == "DOWN"), " (",signif(100*mean(plot.data$dir.values == "DOWN"),sig.figs),"%)", " : ", "Upregulated Genes = ", sum(plot.data$dir.values == "UP"), " (",signif(100*mean(plot.data$dir.values == "UP"),sig.figs),"%)", sep = "")
cur.x.lab <- "Log-2-Fold-Change"
cur.y.lab <- "-log10(p-value)"
cur.colour.lab <- "Direction (FDR < 0.05)"

x.step <- 1
x.max <- x.step*ceiling(max(abs(plot.data$x.values))/x.step)
x.min <- -1*x.max
x.breaks <- seq(from = x.min, to = x.max, by = x.step)

y.step <- 10
y.min <- 0
y.max <- y.step*ceiling(max(plot.data$y.values)/y.step)
y.breaks <- seq(from = y.min, to = y.max, by = y.step)

ggplot(plot.data) + geom_hline(colour = "black", yintercept = 0, lwd = 1) + geom_vline(colour = "black", xintercept = 0, lwd = 1) + geom_point(aes(x = x.values, y = y.values, colour = dir.values), shape = 21, size = 2, fill = "white", alpha = 1, stroke = 1) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + scale_colour_manual(values = colour.value.vec) + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# load in the ARACNe-inferred transcriptional regulatory interactions for TFs, coTFs, and SigPs
tf.network.data <- readRDS(file = "/Volumes/ATG_4TB_HD_v3/ATG_NARNEA_Project/110420/vignette_data/TCGA_LUAD_TF_ARACNe_100_Fold_75%_MCCV_Network_Data.rds")

cotf.network.data <- readRDS(file = "/Volumes/ATG_4TB_HD_v3/ATG_NARNEA_Project/110420/vignette_data/TCGA_LUAD_coTF_ARACNe_100_Fold_75%_MCCV_Network_Data.rds")

sig.network.data <- readRDS(file = "/Volumes/ATG_4TB_HD_v3/ATG_NARNEA_Project/110420/vignette_data/TCGA_LUAD_Sig_ARACNe_100_Fold_75%_MCCV_Network_Data.rds")

total.network.data <- rbind(tf.network.data,cotf.network.data,sig.network.data)

# compute the Regulation Confidence and Mode of Regulation for each interaction
total.network.data$rc.value <- (total.network.data$mi.value*total.network.data$count.value)/max(total.network.data$count.value)
total.network.data$mor.value <- (total.network.data$scc.value)

# create a list of regulon gene sets, one for each of the transcriptional regulatory proteins
total.regulator.names <- unique(as.character(total.network.data$regulator.value))
total.target.names <- unique(as.character(total.network.data$target.value))

true.regulon.list <- lapply(total.regulator.names,function(cur.regulator.name){
	cur.network.data <- total.network.data[which(total.network.data$regulator.value == cur.regulator.name),]
	cur.regulon <- list(tfmode = cur.network.data$mor.value, likelihood = cur.network.data$rc.value)
	names(cur.regulon$tfmode) <- as.character(cur.network.data$target.value)
	return(cur.regulon)
})
names(true.regulon.list) <- total.regulator.names

# keep regulon gene sets with at least 50 targets
true.regulon.list <- true.regulon.list[as.numeric(which(sapply(true.regulon.list,function(x){return(length(x$tfmode))}) >= 50))]

length(true.regulon.list)
length(unlist(lapply(true.regulon.list,function(x){return(names(x$tfmode))})))

# subset to a maximum number of targets per regulator
cur.max.regulon.size <- 500

true.regulon.list <- lapply(true.regulon.list,function(x){
	y <- x$likelihood
	names(y) <- names(x$tfmode)
	y <- names(sort(y, decreasing = TRUE)[seq(from = 1, to = min(cur.max.regulon.size,length(y)), by = 1)])
	z <- list(tfmode = x$tfmode[match(y,names(x$tfmode))], likelihood = x$likelihood[match(y,names(x$tfmode))])
	return(z)
})

# compute the number of regulators, number of targets, and total number of interactions
length(true.regulon.list)

length(unique(unlist(lapply(true.regulon.list,function(x){return(names(x$tfmode))}))))

length(unlist(lapply(true.regulon.list,function(x){return(names(x$tfmode))})))

# create a real gene expression signature from the differential gene expression results
real.luad.vs.lung.ges <- qnorm(p = (cur.dge.res$pvalue/2), lower.tail = FALSE)*sign(cur.dge.res$log2FoldChange)
names(real.luad.vs.lung.ges) <- rownames(cur.dge.res)

# compute the enrichment of all true regulons in the true LUAD vs. LUNG gene expression signature
true.real.dpa.res <- t(sapply(true.regulon.list,function(cur.regulon,cur.ges){
	y <- unlist(NaRnEA(ges = cur.ges, regulon = cur.regulon, minsize = 30, seed = 1, leading.edge = FALSE))
	return(y)
}, cur.ges = real.luad.vs.lung.ges))
true.real.dpa.res <- as.data.frame(true.real.dpa.res)

# create a shuffle gene expression signature
set.seed(sim.seed)
shuffle.luad.vs.lung.ges <- real.luad.vs.lung.ges
names(shuffle.luad.vs.lung.ges) <- sample(names(shuffle.luad.vs.lung.ges), replace = FALSE)

# compute the enrichment of all true regulons in the shuffled LUAD vs. LUNG gene expression signature
true.shuffle.dpa.res <- t(sapply(true.regulon.list,function(cur.regulon,cur.ges){
	y <- unlist(NaRnEA(ges = cur.ges, regulon = cur.regulon, minsize = 30, seed = 1, leading.edge = FALSE))
	return(y)
}, cur.ges = shuffle.luad.vs.lung.ges))
true.shuffle.dpa.res <- as.data.frame(true.shuffle.dpa.res)

# create a synthetic regulon list where the targets of each original regulon are swapped out for targets chosen from the regulon gene set's complement
set.seed(sim.seed)
null.regulon.list <- lapply(true.regulon.list,function(x){
	y <- x
	names(y$tfmode) <- sample(setdiff(names(real.luad.vs.lung.ges),names(x$tfmode)), size = length(y$tfmode))
	return(y)
})

# compute the enrichment of all null regulons in the true LUAD vs. LUNG gene expression signature
null.real.dpa.res <- t(sapply(null.regulon.list,function(cur.regulon,cur.ges){
	y <- unlist(NaRnEA(ges = cur.ges, regulon = cur.regulon, minsize = 30, seed = 1, leading.edge = FALSE))
	return(y)
}, cur.ges = real.luad.vs.lung.ges))
null.real.dpa.res <- as.data.frame(null.real.dpa.res)

# visualize the differential protein activity inferred for the true regulons in the real LUAD vs. LUNG gene expression signature
plot.data <- data.frame(x.values = true.real.dpa.res$pes, y.values = abs(true.real.dpa.res$nes))
plot.data$p.values <- 2*pnorm(q = plot.data$y.values, lower.tail = FALSE)
plot.data$fdr.values <- p.adjust(p = plot.data$p.values, method = "BH")
plot.data$dir.values <- "NONE"
plot.data[which(plot.data$x.values < 0 & plot.data$fdr.values < .05),"dir.values"] <- "DOWN"
plot.data[which(plot.data$x.values > 0 & plot.data$fdr.values < .05),"dir.values"] <- "UP"
plot.data$dir.values <- factor(plot.data$dir.values, levels = c("DOWN","NONE","UP"))

colour.value.vec <- c("DOWN" = "blue1", "NONE" = "grey50", "UP" = "red1")

x.breaks <- seq(from = -1, to = 1, by = .2)

y.step <- 2
y.min <- 0
y.max <- y.step*ceiling(max(plot.data$y.values)/y.step)
y.breaks <- seq(from = y.min, to = y.max, by = y.step)

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("NaRnEA-Inferred Differential Protein Activity","\n","Regulon Gene Sets = Real"," : ","Gene Expression Signature = Real","\n","Deactivated Proteins = ", sum(plot.data$dir.values == "DOWN"), " (",signif(100*mean(plot.data$dir.values == "DOWN"),sig.figs),"%)", " : ", "Activated Proteins = ", sum(plot.data$dir.values == "UP"), " (",signif(100*mean(plot.data$dir.values == "UP"),sig.figs),"%)", sep = "")
cur.x.lab <- "Proportional Enrichment Score"
cur.y.lab <- "| Normalized Enrichment Score |"
cur.colour.lab <- "Direction (FDR < 0.05)"

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = 0) + geom_hline(colour = "black", lwd = 1, yintercept = 0) + geom_point(aes(x = x.values, y = y.values, colour = dir.values), fill = "white", alpha = 1, stroke = 1.5, size = 2, shape = 21) + scale_colour_manual(values = colour.value.vec) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# visualize the differential protein activity inferred for the true regulons in the shuffled LUAD vs. LUNG gene expression signature
plot.data <- data.frame(x.values = true.shuffle.dpa.res$pes, y.values = abs(true.shuffle.dpa.res$nes))
plot.data$p.values <- 2*pnorm(q = plot.data$y.values, lower.tail = FALSE)
plot.data$fdr.values <- p.adjust(p = plot.data$p.values, method = "BH")
plot.data$dir.values <- "NONE"
plot.data[which(plot.data$x.values < 0 & plot.data$fdr.values < .05),"dir.values"] <- "DOWN"
plot.data[which(plot.data$x.values > 0 & plot.data$fdr.values < .05),"dir.values"] <- "UP"
plot.data$dir.values <- factor(plot.data$dir.values, levels = c("DOWN","NONE","UP"))

colour.value.vec <- c("DOWN" = "blue1", "NONE" = "grey50", "UP" = "red1")

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("NaRnEA-Inferred Differential Protein Activity","\n","Regulon Gene Sets = Real"," : ","Gene Expression Signature = Shuffled","\n","Deactivated Proteins = ", sum(plot.data$dir.values == "DOWN"), " (",signif(100*mean(plot.data$dir.values == "DOWN"),sig.figs),"%)", " : ", "Activated Proteins = ", sum(plot.data$dir.values == "UP"), " (",signif(100*mean(plot.data$dir.values == "UP"),sig.figs),"%)", sep = "")
cur.x.lab <- "Proportional Enrichment Score"
cur.y.lab <- "| Normalized Enrichment Score |"
cur.colour.lab <- "Direction (FDR < 0.05)"

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = 0) + geom_hline(colour = "black", lwd = 1, yintercept = 0) + geom_point(aes(x = x.values, y = y.values, colour = dir.values), fill = "white", alpha = 1, stroke = 1.5, size = 2, shape = 21) + scale_colour_manual(values = colour.value.vec) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# visualize the differential protein activity inferred for the shuffled regulons in the real LUAD vs. LUNG gene expression signature
plot.data <- data.frame(x.values = null.real.dpa.res$pes, y.values = abs(null.real.dpa.res$nes))
plot.data$p.values <- 2*pnorm(q = plot.data$y.values, lower.tail = FALSE)
plot.data$fdr.values <- p.adjust(p = plot.data$p.values, method = "BH")
plot.data$dir.values <- "NONE"
plot.data[which(plot.data$x.values < 0 & plot.data$fdr.values < .05),"dir.values"] <- "DOWN"
plot.data[which(plot.data$x.values > 0 & plot.data$fdr.values < .05),"dir.values"] <- "UP"
plot.data$dir.values <- factor(plot.data$dir.values, levels = c("DOWN","NONE","UP"))

colour.value.vec <- c("DOWN" = "blue1", "NONE" = "grey50", "UP" = "red1")

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("NaRnEA-Inferred Differential Protein Activity","\n","Regulon Gene Sets = Shuffled"," : ","Gene Expression Signature = Real","\n","Deactivated Proteins = ", sum(plot.data$dir.values == "DOWN"), " (",signif(100*mean(plot.data$dir.values == "DOWN"),sig.figs),"%)", " : ", "Activated Proteins = ", sum(plot.data$dir.values == "UP"), " (",signif(100*mean(plot.data$dir.values == "UP"),sig.figs),"%)", sep = "")
cur.x.lab <- "Proportional Enrichment Score"
cur.y.lab <- "| Normalized Enrichment Score |"
cur.colour.lab <- "Direction (FDR < 0.05)"

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = 0) + geom_hline(colour = "black", lwd = 1, yintercept = 0) + geom_point(aes(x = x.values, y = y.values, colour = dir.values), fill = "white", alpha = 1, stroke = 1.5, size = 2, shape = 21) + scale_colour_manual(values = colour.value.vec) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# visualize the empirical cumulative distribution of two-sided p-values for (1) the real regulon gene sets in the real LUAD vs. LUNG gene expression signature, (2) the real regulon gene sets in the shuffled LUAD vs. LUNG gene expression signature, and (3) the shuffled regulon gene sets in the real LUAD vs. LUNG gene expression signature 
plot.data.true.real <- data.frame(x.values = seq(from = 0, to = 1, length.out = 1E3))
plot.data.true.real$y.values <- ecdf((2*pnorm(q = abs(true.real.dpa.res$nes), lower.tail = FALSE)))(plot.data.true.real$x.values)
plot.data.true.real$group.values <- "Real-Real"

plot.data.true.shuffle <- data.frame(x.values = seq(from = 0, to = 1, length.out = 1E3))
plot.data.true.shuffle$y.values <- ecdf((2*pnorm(q = abs(true.shuffle.dpa.res$nes), lower.tail = FALSE)))(plot.data.true.shuffle$x.values)
plot.data.true.shuffle$group.values <- "Real-Shuffled"

plot.data.null.real <- data.frame(x.values = seq(from = 0, to = 1, length.out = 1E3))
plot.data.null.real$y.values <- ecdf((2*pnorm(q = abs(null.real.dpa.res$nes), lower.tail = FALSE)))(plot.data.null.real$x.values)
plot.data.null.real$group.values <- "Shuffled-Real"

plot.data <- rbind(plot.data.true.real, plot.data.true.shuffle, plot.data.null.real)
plot.data$group.values <- factor(plot.data$group.values, levels = unique(as.character(plot.data$group.values)))

colour.value.vec <- c("Real-Real" = "red1", "Real-Shuffled" = "blue1", "Shuffled-Real" = "forestgreen")

x.breaks <- seq(from = 0, to = 1, by = .1)
y.breaks <- x.breaks

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("NaRnEA-Inferred Differential Protein Activity", "\n", "Real-Real", " : ", "p < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(true.real.dpa.res$nes), lower.tail = FALSE)), method = "none") < .05)), sig.figs), "%", " : ", "FDR < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(true.real.dpa.res$nes), lower.tail = FALSE)), method = "BH") < .05)), sig.figs), "%", " : ", "FWER < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(true.real.dpa.res$nes), lower.tail = FALSE)), method = "bonferroni") < .05)), sig.figs), "%", "\n", "Real-Shuffled", " : ", "p < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(true.shuffle.dpa.res$nes), lower.tail = FALSE)), method = "none") < .05)), sig.figs), "%", " : ", "FDR < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(true.shuffle.dpa.res$nes), lower.tail = FALSE)), method = "BH") < .05)), sig.figs), "%", " : ", "FWER < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(true.shuffle.dpa.res$nes), lower.tail = FALSE)), method = "bonferroni") < .05)), sig.figs), "%", "\n", "Shuffled-Real", " : ", "p < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(null.real.dpa.res$nes), lower.tail = FALSE)), method = "none") < .05)), sig.figs), "%", " : ", "FDR < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(null.real.dpa.res$nes), lower.tail = FALSE)), method = "BH") < .05)), sig.figs), "%", " : ", "FWER < 0.05 = ", signif((100*mean(p.adjust(p = (2*pnorm(q = abs(null.real.dpa.res$nes), lower.tail = FALSE)), method = "bonferroni") < .05)), sig.figs), "%", sep = "")
cur.x.lab <- "NaRnEA Two-Sided P-Value"
cur.y.lab <- "Empirical Cumulative Probability"
cur.colour.lab <- "Regulon-GES Combo"

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = range(x.breaks)) + geom_hline(colour = "black", lwd = 1, yintercept = range(y.breaks)) + geom_abline(colour = "black", lwd = 1, intercept = 0, slope = 1, linetype = "dashed") + geom_line(aes(x = x.values, y = y.values, colour = group.values), lwd = 1.5) + scale_colour_manual(values = colour.value.vec) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# visualize the enrichment of the regulon gene set for FOXM1 in the LUAD vs. LUNG gene expression signature with an approximate 95% confidence interval inferred with the bootstrap percentile method
set.seed(sim.seed)

cur.hugo.name <- "FOXM1"

cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")

cur.regulon <- true.regulon.list[[cur.entrez.name]]

boot.num <- 1000

cur.boot.idx.list <- lapply(1:boot.num,function(i){
	y <- sample(1:length(cur.regulon$tfmode), size = length(cur.regulon$tfmode), replace = TRUE)
	return(y)
})

cur.enrichment.res <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = cur.regulon)

cur.boot.pes.values <- sapply(cur.boot.idx.list,function(sub.idx, cur.ges){
	sub.regulon <- list(tfmode = cur.regulon$tfmode[sub.idx], likelihood = cur.regulon$likelihood[sub.idx])
	sub.pes <- NaRnEA(ges = cur.ges, regulon = sub.regulon)$pes
	return(sub.pes)
}, cur.ges = real.luad.vs.lung.ges)

cur.title <- paste("NaRnEA-Inferred Differential Protein Activity", "\n", "Regulator = ", cur.hugo.name, " : ", "Targets = ", length(cur.regulon$tfmode), sep = "")
cur.subtitle <- paste("NaRnEA"," : ", "PES = ", signif(cur.enrichment.res$pes, sig.figs), " [", signif(as.numeric(quantile(cur.boot.pes.values,probs = c(.025))), sig.figs), ",", signif(as.numeric(quantile(cur.boot.pes.values,probs = c(.975))), sig.figs),"]", " : ", "NES = ", signif(cur.enrichment.res$nes, sig.figs), " : ", "p = ", signif((2*pnorm(q = abs(cur.enrichment.res$nes), lower.tail = FALSE)), sig.figs), sep = "")
cur.x.lab <- "LUAD vs. LUNG Gene Expression Signature"

EnrichmentAnalysisPlot(cur.ges = real.luad.vs.lung.ges, cur.regul = true.regulon.list[[cur.entrez.name]], cur.title = cur.title, cur.subtitle = cur.subtitle, cur.x.lab = cur.x.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")

# visualize the enrichment of the regulon gene set for CENPF in the LUAD vs. LUNG gene expression signature with an approximate 95% confidence interval inferred with the bootstrap percentile method
set.seed(sim.seed)

cur.hugo.name <- "CENPF"

cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")

cur.regulon <- true.regulon.list[[cur.entrez.name]]

boot.num <- 1000

cur.boot.idx.list <- lapply(1:boot.num,function(i){
	y <- sample(1:length(cur.regulon$tfmode), size = length(cur.regulon$tfmode), replace = TRUE)
	return(y)
})

cur.enrichment.res <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = cur.regulon)

cur.boot.pes.values <- sapply(cur.boot.idx.list,function(sub.idx, cur.ges){
	sub.regulon <- list(tfmode = cur.regulon$tfmode[sub.idx], likelihood = cur.regulon$likelihood[sub.idx])
	sub.pes <- NaRnEA(ges = cur.ges, regulon = sub.regulon)$pes
	return(sub.pes)
}, cur.ges = real.luad.vs.lung.ges)

cur.title <- paste("NaRnEA-Inferred Differential Protein Activity", "\n", "Regulator = ", cur.hugo.name, " : ", "Targets = ", length(cur.regulon$tfmode), sep = "")
cur.subtitle <- paste("NaRnEA"," : ", "PES = ", signif(cur.enrichment.res$pes, sig.figs), " [", signif(as.numeric(quantile(cur.boot.pes.values,probs = c(.025))), sig.figs), ",", signif(as.numeric(quantile(cur.boot.pes.values,probs = c(.975))), sig.figs),"]", " : ", "NES = ", signif(cur.enrichment.res$nes, sig.figs), " : ", "p = ", signif((2*pnorm(q = abs(cur.enrichment.res$nes), lower.tail = FALSE)), sig.figs), sep = "")
cur.x.lab <- "LUAD vs. LUNG Gene Expression Signature"

EnrichmentAnalysisPlot(cur.ges = real.luad.vs.lung.ges, cur.regul = true.regulon.list[[cur.entrez.name]], cur.title = cur.title, cur.subtitle = cur.subtitle, cur.x.lab = cur.x.lab)

ggsave(filename = paste("Fig_",figure.output.value,"_",(1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_",figure.output.value,"_",sep = "")))),".png",sep = ""), path = "PNG_Figures", dpi = 500, device = "png")


## Permutation Test for Equality of FOXM1 and CENFP regulon gene set enrichment

# assign the FOXM1 regulon gene set as gene set a and the CENPF regulon gene set as gene set b
cur.hugo.name <- "FOXM1"
cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")
synthetic.gene.set.a <- true.regulon.list[[cur.entrez.name]]
gene.set.a.size <- length(synthetic.gene.set.a$tfmode)

cur.hugo.name <- "CENPF"
cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")
synthetic.gene.set.b <- true.regulon.list[[cur.entrez.name]]
gene.set.b.size <- length(synthetic.gene.set.b$tfmode)

# check to see how much the regulon gene sets overlap
length(intersect(names(synthetic.gene.set.a$tfmode),names(synthetic.gene.set.b$tfmode)))

# combine the gene sets to create a single gene set
synthetic.gene.set.c <- list(tfmode = c(synthetic.gene.set.a$tfmode, synthetic.gene.set.b$tfmode), likelihood = c(synthetic.gene.set.a$likelihood, synthetic.gene.set.b$likelihood))

# create permutations of the synthetic gene sets
set.seed(sim.seed)

gene.set.a.per <- 5000

synthetic.gene.set.a.null.idx.list <- lapply(1:gene.set.a.per,function(i){
	y <- sample(1:length(synthetic.gene.set.c$tfmode), size = gene.set.a.size, replace = FALSE)
	return(y)
})
names(synthetic.gene.set.a.null.idx.list) <- paste("na",1:length(synthetic.gene.set.a.null.idx.list),"",sep = "_")

synthetic.gene.set.b.null.idx.list <- lapply(1:gene.set.a.per,function(i){
	y <- setdiff(1:length(synthetic.gene.set.c$tfmode), synthetic.gene.set.a.null.idx.list[[i]])
	return(y)
})
names(synthetic.gene.set.b.null.idx.list) <- paste("nb",1:length(synthetic.gene.set.b.null.idx.list),"",sep = "_")

# compute the enrichment of the original gene sets and the permuted gene sets in the real LUAD vs. LUNG gene expression signature
synthetic.gene.set.a.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.a)

synthetic.gene.set.a.null.enrichment <- t(sapply(synthetic.gene.set.a.null.idx.list,function(sub.idx){
	sub.gene.set <- list(tfmode = synthetic.gene.set.c$tfmode[sub.idx], likelihood = synthetic.gene.set.c$likelihood[sub.idx])
	sub.res <- unlist(NaRnEA(ges = real.luad.vs.lung.ges, regulon = sub.gene.set))
	return(sub.res)
}))

synthetic.gene.set.b.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.b)

synthetic.gene.set.b.null.enrichment <- t(sapply(synthetic.gene.set.b.null.idx.list,function(sub.idx){
	sub.gene.set <- list(tfmode = synthetic.gene.set.c$tfmode[sub.idx], likelihood = synthetic.gene.set.c$likelihood[sub.idx])
	sub.res <- unlist(NaRnEA(ges = real.luad.vs.lung.ges, regulon = sub.gene.set))
	return(sub.res)
}))

# visualize the results
plot.data <- data.frame(x.values = abs(synthetic.gene.set.a.null.enrichment[,"pes"] - synthetic.gene.set.b.null.enrichment[,"pes"]))

x.min <- 0
x.max <- 2
x.step <- .2
x.breaks <- seq(from = x.min, to = x.max, by = x.step)

bin.prop <- .1
bin.breaks <- seq(from = x.min, to = x.max, by = (x.step*bin.prop))

y.step <- 2
y.min <- 0
y.max <- y.step*ceiling(max(density(plot.data$x.values, from = 0, to = 2)$y)/y.step)
y.breaks <- seq(from = y.min, to = y.max, by = y.step)

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("FOXM1"," : ","PES = ",signif(synthetic.gene.set.a.enrichment$pes,sig.figs)," : ","NES = ", signif(synthetic.gene.set.a.enrichment$nes, sig.figs)," (p = ",signif((2*pnorm(q = abs(synthetic.gene.set.a.enrichment$nes), lower.tail = FALSE)),sig.figs),")","\n", "CENPF"," : ","PES = ",signif(synthetic.gene.set.b.enrichment$pes,sig.figs)," : ","NES = ", signif(synthetic.gene.set.b.enrichment$nes, sig.figs)," (p = ",signif((2*pnorm(q = abs(synthetic.gene.set.b.enrichment$nes), lower.tail = FALSE)),sig.figs),")", "\n", "Empirical Null Distribution Permutations = ", gene.set.a.per, "\n", "| (FOXM1 PES)  - (CENPF PES) | = ", signif(abs(synthetic.gene.set.a.enrichment$pes - synthetic.gene.set.b.enrichment$pes), sig.figs), " (p = ",signif((sum(plot.data$x.values >= abs(synthetic.gene.set.a.enrichment$pes - synthetic.gene.set.b.enrichment$pes)) + 1)/(nrow(plot.data) + 1), sig.figs),")",sep = "")
cur.x.lab <- "| (Null FOXM1 PES)  - (Null CENPF PES) |"
cur.y.lab <- "Probability Density"

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = range(x.breaks)) + geom_density(aes(x = x.values, y = ..density..), colour = "red1", fill = "red1", alpha = .25, lwd = 1) + geom_vline(colour = "red1", lwd = 1.5, xintercept = abs(synthetic.gene.set.a.enrichment$pes - synthetic.gene.set.b.enrichment$pes), linetype = "dashed") + geom_hline(colour = "black", lwd = 1, yintercept = 0) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18)) + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab)

ggsave(filename = paste("Fig_", figure.output.value, "_", (1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_", figure.output.value, "_", sep = "")))), ".png", sep = ""), device = "png", path = "PNG_Figures", dpi = 500)



## Permutation Test for Interaction between the FOXM1 regulon gene set and the CENPF regulon gene set

# assign the FOXM1 regulon gene set as gene set a and the CENPF regulon gene set as gene set b
cur.hugo.name <- "FOXM1"
cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")
synthetic.gene.set.a <- true.regulon.list[[cur.entrez.name]]
gene.set.a.size <- length(synthetic.gene.set.a$tfmode)

cur.hugo.name <- "CENPF"
cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")
synthetic.gene.set.b <- true.regulon.list[[cur.entrez.name]]
gene.set.b.size <- length(synthetic.gene.set.b$tfmode)

# identify target genes common to both regulon gene sets
common.targets <- (intersect(names(synthetic.gene.set.a$tfmode),names(synthetic.gene.set.b$tfmode)))

# create synthetic gene set c from the targets in synthetic gene set a (FOXM1) which are not present in synthetic gene set b (CENPF)
synthetic.gene.set.c <- list(tfmode = synthetic.gene.set.a$tfmode[which(!(names(synthetic.gene.set.a$tfmode)%in%common.targets))], likelihood = synthetic.gene.set.a$likelihood[which(!(names(synthetic.gene.set.a$tfmode)%in%common.targets))])
gene.set.c.size <- length(synthetic.gene.set.c$tfmode)

# create synthetic gene set d from the targets in synthetic gene set b (CENPF) which are not present in synthetic gene set a (FOXM1)
synthetic.gene.set.d <- list(tfmode = synthetic.gene.set.b$tfmode[which(!(names(synthetic.gene.set.b$tfmode)%in%common.targets))], likelihood = synthetic.gene.set.b$likelihood[which(!(names(synthetic.gene.set.b$tfmode)%in%common.targets))])
gene.set.d.size <- length(synthetic.gene.set.d$tfmode)

# compute the empirical null distribution for the enrichment of gene set c in the gene expression signature using permutations
set.seed(sim.seed)

gene.set.c.per <- 5000

synthetic.gene.set.c.null.idx.list <- lapply(1:gene.set.c.per,function(i){
	y <- sample(1:length(synthetic.gene.set.a$tfmode), size = gene.set.c.size, replace = FALSE)
	return(y)
})
names(synthetic.gene.set.c.null.idx.list) <- paste("nc",1:length(synthetic.gene.set.c.null.idx.list),"",sep = "_")

synthetic.gene.set.c.null.enrichment <- t(sapply(synthetic.gene.set.c.null.idx.list,function(sub.idx){
	sub.gene.set <- list(tfmode = synthetic.gene.set.a$tfmode[sub.idx], likelihood = synthetic.gene.set.a$likelihood[sub.idx])
	sub.res <- unlist(NaRnEA(ges = real.luad.vs.lung.ges, regulon = sub.gene.set))
	return(sub.res)
}))

synthetic.gene.set.a.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.a)

synthetic.gene.set.c.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.c)

synthetic.gene.set.c.q.value <- (sum(synthetic.gene.set.c.null.enrichment[,"pes"] < synthetic.gene.set.c.enrichment$pes) + 1)/(gene.set.c.per + 2)
synthetic.gene.set.c.z.value <- qnorm(p = synthetic.gene.set.c.q.value, lower.tail = TRUE)
synthetic.gene.set.c.p.value <- 2*pnorm(q = abs(synthetic.gene.set.c.z.value), lower.tail = FALSE)

# compute the empirical null distribution for the enrichment of gene set d in the gene expression signature using permutations
set.seed(sim.seed)

gene.set.d.per <- gene.set.c.per

synthetic.gene.set.d.null.idx.list <- lapply(1:gene.set.d.per,function(i){
	y <- sample(1:length(synthetic.gene.set.b$tfmode), size = gene.set.d.size, replace = FALSE)
	return(y)
})
names(synthetic.gene.set.d.null.idx.list) <- paste("nc",1:length(synthetic.gene.set.d.null.idx.list),"",sep = "_")

synthetic.gene.set.d.null.enrichment <- t(sapply(synthetic.gene.set.d.null.idx.list,function(sub.idx){
	sub.gene.set <- list(tfmode = synthetic.gene.set.b$tfmode[sub.idx], likelihood = synthetic.gene.set.b$likelihood[sub.idx])
	sub.res <- unlist(NaRnEA(ges = real.luad.vs.lung.ges, regulon = sub.gene.set))
	return(sub.res)
}))

synthetic.gene.set.b.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.b)

synthetic.gene.set.d.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.d)

synthetic.gene.set.d.q.value <- (sum(synthetic.gene.set.d.null.enrichment[,"pes"] < synthetic.gene.set.d.enrichment$pes) + 1)/(gene.set.d.per + 2)
synthetic.gene.set.d.z.value <- qnorm(p = synthetic.gene.set.d.q.value, lower.tail = TRUE)
synthetic.gene.set.d.p.value <- 2*pnorm(q = abs(synthetic.gene.set.d.z.value), lower.tail = FALSE)

# visualize the results
plot.data <- data.frame(x.values = c(synthetic.gene.set.c.null.enrichment[,"pes"], synthetic.gene.set.d.null.enrichment[,"pes"]), group.values = c(rep("FOXM1 Diff", times = gene.set.c.per), rep("CENPF Diff", times = gene.set.d.per)))

plot.data$group.values <- factor(plot.data$group.values, levels = c("FOXM1 Diff","CENPF Diff"))

colour.value.vec <- c("FOXM1 Diff" = "red1", "CENPF Diff" = "blue1")
fill.value.vec <- colour.value.vec

x.min <- -1
x.max <- 1
x.step <- .2
x.breaks <- seq(from = x.min, to = x.max, by = x.step)

y.step <- 3
y.min <- 0
y.max <- y.step*ceiling(max(c(max(density(plot.data[which(plot.data$group.values == "FOXM1 Diff"),"x.values"], from = -1, to = 1)$y),max(density(plot.data[which(plot.data$group.values == "CENPF Diff"),"x.values"], from = -1, to = 1)$y)))/y.step)
y.breaks <- seq(from = y.min, to = y.max, by = y.step)

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("FOXM1"," : ","PES = ", signif(synthetic.gene.set.a.enrichment$pes,sig.figs), " : ", "NES = ", signif(synthetic.gene.set.a.enrichment$nes, sig.figs), " (p = ",signif((2*pnorm(q = abs(synthetic.gene.set.a.enrichment$nes), lower.tail = FALSE)),sig.figs),")", "\n", "FOXM1 Diff"," : ","PES = ", signif(synthetic.gene.set.c.enrichment$pes,sig.figs), " : ", "Q = ", signif(synthetic.gene.set.c.q.value, sig.figs), " : ", "Z = ", signif(synthetic.gene.set.c.z.value,sig.figs), " (p = ",signif(synthetic.gene.set.c.p.value,sig.figs),")", "\n", "CENPF"," : ","PES = ", signif(synthetic.gene.set.b.enrichment$pes,sig.figs), " : ", "NES = ", signif(synthetic.gene.set.b.enrichment$nes, sig.figs), " (p = ",signif((2*pnorm(q = abs(synthetic.gene.set.b.enrichment$nes), lower.tail = FALSE)),sig.figs),")", "\n", "CENPF Diff"," : ","PES = ", signif(synthetic.gene.set.d.enrichment$pes,sig.figs), " : ", "Q = ", signif(synthetic.gene.set.d.q.value, sig.figs), " : ", "Z = ", signif(synthetic.gene.set.d.z.value,sig.figs), " (p = ",signif(synthetic.gene.set.d.p.value,sig.figs),")",sep = "")
cur.x.lab <- "Proportional Enrichment Score"
cur.y.lab <- "Probability Density"
cur.colour.lab <- "Regulon Gene Set"
cur.fill.lab <- cur.colour.lab

ggplot(plot.data) + geom_vline(colour = "black", lwd = 1, xintercept = 0) + geom_density(aes(x = x.values, y = ..density.., colour = group.values, fill = group.values), alpha = .25, lwd = 1) + geom_hline(colour = "black", lwd = 1, yintercept = 0) + scale_x_continuous(limits = range(x.breaks), breaks = x.breaks) + scale_y_continuous(limits = range(y.breaks), breaks = y.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.x = element_text(hjust = 0.5, colour = "black", size = 18), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom") + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab, fill = cur.fill.lab) + scale_colour_manual(values = colour.value.vec) + scale_fill_manual(values = fill.value.vec) + geom_vline(colour = "red1", lwd = 1, linetype = "dashed", xintercept = synthetic.gene.set.c.enrichment$pes) + geom_vline(colour = "blue1", lwd = 1, linetype = "dashed", xintercept = synthetic.gene.set.d.enrichment$pes)

ggsave(filename = paste("Fig_", figure.output.value, "_", (1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_", figure.output.value, "_", sep = "")))), ".png", sep = ""), device = "png", path = "PNG_Figures", dpi = 500)


# Leading Edge Analysis for FOXM1 and CENPF

# assign the FOXM1 regulon gene set as gene set a and the CENPF regulon gene set as gene set b
cur.hugo.name <- "FOXM1"
cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")
synthetic.gene.set.a <- true.regulon.list[[cur.entrez.name]]
gene.set.a.size <- length(synthetic.gene.set.a$tfmode)

cur.hugo.name <- "CENPF"
cur.entrez.name <- paste("g",names(sort(table(entrez.to.hugo.data[which(entrez.to.hugo.data$gene == cur.hugo.name),"entrez"]), decreasing = TRUE))[1],"",sep = "_")
synthetic.gene.set.b <- true.regulon.list[[cur.entrez.name]]
gene.set.b.size <- length(synthetic.gene.set.b$tfmode)

# compute the enrichment of each regulon gene set in the LUAD vs. LUNG gene expression signature with leading edge analysis
gene.set.a.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.a, leading.edge = TRUE)

gene.set.b.enrichment <- NaRnEA(ges = real.luad.vs.lung.ges, regulon = synthetic.gene.set.b, leading.edge = TRUE)

# visualize the leading edge enrichment results 
total.target.names <- unique(c(names(gene.set.a.enrichment$ledge),names(gene.set.b.enrichment$ledge)))
common.target.names <- intersect(names(gene.set.a.enrichment$ledge),names(gene.set.b.enrichment$ledge))

plot.data <- data.frame(x.values = c(gene.set.a.enrichment$ledge,  gene.set.b.enrichment$ledge), gene.values = c(names(gene.set.a.enrichment$ledge),names(gene.set.b.enrichment$ledge)), set.values = c(rep("FOXM1", times = length(gene.set.a.enrichment$ledge)), rep("CENPF", times = length(gene.set.b.enrichment$ledge))))
plot.data$group.values <- paste(plot.data$set.values,"ONLY", sep = " ")
plot.data[which(plot.data$gene.values%in%common.target.names), "group.values"] <- "FOXM1 & CENPF"
plot.data$set.values <- factor(plot.data$set.values, levels = c("FOXM1","CENPF"))
plot.data$group.values <- factor(plot.data$group.values, levels = c("FOXM1 ONLY","CENPF ONLY","FOXM1 & CENPF"))
plot.data <- plot.data[order(plot.data$set.values, plot.data$x.values, decreasing = FALSE),]
plot.data$y.values <- 1:nrow(plot.data)
plot.data$z.values <- qnorm(p = plot.data$x.values, lower.tail = FALSE)

colour.value.vec <- c("FOXM1 ONLY" = "red1", "CENPF ONLY" = "blue1", "FOXM1 & CENPF" = "grey50")

z.step <- 1
z.max <- z.step*ceiling(max(abs(plot.data$z.values))/z.step)
z.min <- -1*z.max
z.breaks <- seq(from = z.min, to = z.max, by = z.step)

gene.set.a.wilcox.res <- wilcox.test(x = plot.data[which(plot.data$group.values == "FOXM1 & CENPF"),"z.values"], y = plot.data[which(plot.data$group.values == "FOXM1 ONLY"),"z.values"], alternative = "two.sided", conf.int = TRUE, paired = FALSE)

gene.set.b.wilcox.res <- wilcox.test(x = plot.data[which(plot.data$group.values == "FOXM1 & CENPF"),"z.values"], y = plot.data[which(plot.data$group.values == "CENPF ONLY"),"z.values"], alternative = "two.sided", conf.int = TRUE, paired = FALSE)

cur.title <- "TCGA LUAD (50) vs. TCGA LUNG (50)"
cur.subtitle <- paste("FOXM1 & CENPF Leading Edge Analysis", "\n", "MWU Test for FOXM1 ONLY vs. FOXM1 & CENPF"," : ", "RBS = ", signif((2*as.numeric(gene.set.a.wilcox.res$statistic)/(sum(plot.data$group.values == "FOXM1 ONLY")*sum(plot.data$group.values == "FOXM1 & CENPF")) - 1), sig.figs), " (p = ", signif(gene.set.a.wilcox.res$p.value, sig.figs),")", "\n", "MWU Test for CENPF ONLY vs. FOXM1 & CENPF"," : ", "RBS = ", signif((2*as.numeric(gene.set.b.wilcox.res$statistic)/(sum(plot.data$group.values == "CENPF ONLY")*sum(plot.data$group.values == "FOXM1 & CENPF")) - 1), sig.figs), " (p = ", signif(gene.set.b.wilcox.res$p.value, sig.figs),")",sep = "")
cur.y.lab <- "Leading Edge Z-Score"
cur.x.lab <- "Regulon Gene Set Member"
cur.colour.lab <- "Regulon Gene Set Membership"

ggplot(plot.data) + geom_hline(colour = "black", lwd = 1, yintercept = 0) + facet_wrap(~set.values, ncol = 1, scales = "free") + geom_segment(aes(x = y.values, xend = y.values, y = z.values, yend = 0, colour = group.values), lwd = .75) + scale_colour_manual(values = colour.value.vec) + scale_y_continuous(limits = range(z.breaks), breaks = z.breaks) + theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 25), plot.subtitle = element_text(hjust = 0.5, colour = "black", size = 22), axis.title = element_text(hjust = 0.5, colour = "black", size = 20), axis.text.y = element_text(hjust = 1, colour = "black", size = 18), axis.text.x = element_blank(), legend.title = element_text(hjust = 0, colour = "black", size = 20), legend.text = element_text(hjust = 0, colour = "black", size = 18), legend.position = "bottom", strip.text = element_text(hjust = 0.5, colour = "black", size = 20), axis.ticks.x = element_blank()) + labs(title = cur.title, subtitle = cur.subtitle, x = cur.x.lab, y = cur.y.lab, colour = cur.colour.lab, fill = cur.fill.lab)

ggsave(filename = paste("Fig_", figure.output.value, "_", (1 + length(list.files(path = "PNG_Figures", pattern = paste("Fig_", figure.output.value, "_", sep = "")))), ".png", sep = ""), device = "png", path = "PNG_Figures", dpi = 500)



```









